README.md
# What time is it Mr. Wolf?

A browser extension that enhances your experience with the Smaregi Timecard system by calculating and displaying the remaining working hours for the week, along with the expected end time for the day. Stay on top of your work schedule and never miss a beat with Smaregi Timecard Helper!

Go to the chrome web store and add the [What time is it Mr. Wolf?](https://chrome.google.com/webstore/detail/what-time-is-it-mrwolf/nckligekkdbanfgnilneakpknpklllge) extension.

### Features

- Automatically calculates the remaining working hours for the week
- Calculates the expected end time for the day, considering breaks
- Displays a convenient summary with the remaining time and end time
- Automatically updates the displayed information when necessary
- Lightweight and unobtrusive, only runs when you're on the Smaregi Timecard website
- Works with Google Chrome and compatible browsers

### Installation
1. Clone this repository or download it as a ZIP file.
2. In Google Chrome, go to the extensions page (chrome://extensions).
3. Enable "Developer mode" by toggling the switch in the top-right corner.
4. Click the "Load unpacked" button and select the directory containing the downloaded repository.
5. The What time is it Mr. Wolf? extension should now appear in your extensions list, and its icon should be visible in the toolbar. 

### Usage
After installing the extension, simply navigate to the Smaregi Timecard website (timecard1.smaregi.jp/staffs/dashboard). The extension will automatically run and display the remaining working hours for the week, along with the expected end time for the day, on the web page.

### Configuration
You can customize the default working hours per week and the default starting time by modifying the corresponding values in the background script (background.js). The default values are 40 hours per week and 8:00 AM as the starting time.

```javascript
const defaultWeeklyHours = 40;
const defaultStartTime = 8;
```

### Contributing
We welcome contributions to improve the What time is it Mr. Wolf? extension. If you'd like to contribute, please fork the repository and submit a pull request with your changes.

### License
What time is it Mr. Wolf? is released under the MIT License.


CHANGELOG.md
# Changelog

## 0.2.9 (2023-11-14)
- 🐞 Fix for number of days worked

## 0.2.8 (2023-04-04)
- 🐞 Handling querySelector bug fix

## 0.2.7 (2023-04-04)
- 🛠️ More accurate selection of days worked and start and end times
- 🎨 Changed day finish message and media query width
## 0.2.6 (2023-04-01)
- 📖 Add docs
- ⭐️ Add mutation observer

## 0.2.5 (2023-03-31)
- 🐞 Handling NaN bug fix

## 0.2.3, 0.2.4 (2023-03-31)
- 🏗️ Remove debug code

## 0.2.2 (2023-03-30)
- 🚦 Adding Jest testing and submit button spinner

## 0.2.1 (2023-03-27)

- 🐞 New fix for infinity bug

## 0.2 (2023-02-28)

- 🛠️ Add weekend message
- 🐞 handle infinity bug

## 0.1 (2023-02-28)

- ⭐️ Initial setup


Mr_Wolf/scripts/index.test.js
/**
 * @jest-environment jsdom
 * @jest-environment-options {"url": "https://jestjs.io/"}
 */

const {
  timeLeft,
  startTime,
  calculateBreak,
  endTime,
  singleDigits,
} = require("./index");

/**
 * Calculate the time left to work this week.
 * @param {number} weeklyHours - The number of hours the user is expected to work in a week.
 * @param {number} workedHours - The number of hours the user has already worked this week.
 * @returns {string} - The formatted time left to work this week.
 */
describe("timeLeft", () => {
  test("Beginning of the week: 40hrs - 0hr", () => {
    expect(timeLeft(40, 0)).toBe(40);
  });

  test("End of the week: 40hrs - 40hr", () => {
    expect(timeLeft(40, 40)).toBe(0);
  });

  test("Overtime: 40hrs - 41hr", () => {
    expect(timeLeft(40, 41)).toBe(0);
  });

  test("Custom weekly hours: 35hrs - 0hr", () => {
    expect(timeLeft(35, 0)).toBe(35);
  });

  test("Custom weekly hours: 45hrs - 45hr", () => {
    expect(timeLeft(45, 45)).toBe(0);
  });
});

/**
 * Get the start time from an array of time strings or use the default start time.
 * @param {string[]|undefined} timeArray - An array containing start time(s) as strings in the "HH:mm" format.
 * @param {number} defaultStart - A default start time as a number (e.g., 8 for 8 AM).
 * @returns {number} - The start time as a number (e.g., 9.25 for 9:15 AM).
 */
describe("startTime", () => {
  test("No Start Time", () => {
    expect(startTime(undefined, 8)).toBe(8);
  });

  test("Only Start Time", () => {
    expect(startTime(["09:00"], 8)).toBe(9);
  });

  // Cannot calculate end time if you have multiple start times. So will be handled later with message.
  test("Multiple Start Times", () => {
    expect(startTime(["09:00", "10:00", "11:00"], 8)).toBe(9);
  });

  test("Start time with leading zero", () => {
    expect(startTime(["07:30"], 8)).toBe(7.5);
  });

  test("Start time with non-zero minutes", () => {
    expect(startTime(["09:15"], 8)).toBe(9.25);
  });

  test("Custom default start time: 7", () => {
    expect(startTime(undefined, 7)).toBe(7);
  });

  test("Custom default start time: 9", () => {
    expect(startTime(undefined, 9)).toBe(9);
  });
});

/**
 * Calculate the total duration of breaks in hours.
 * @param {string[]|undefined} breakArray - An array containing break time(s) as strings in the "HH:mm" format.
 * @returns {number} - The total duration of breaks in hours.
 */
describe("calculateBreak", () => {
  test("No break", () => {
    expect(calculateBreak(undefined)).toBe(1);
  });

  test("One break", () => {
    expect(calculateBreak(["12:00", "13:00"])).toBe(1);
  });

  test("Multiple breaks, odd", () => {
    expect(calculateBreak(["12:00", "13:00", "17:00"])).toBe(1);
  });

  test("Multiple breaks, even", () => {
    expect(calculateBreak(["12:00", "13:00", "17:00", "18:00"])).toBe(2.0);
  });
  test("Multiple breaks, even", () => {
    expect(
      calculateBreak(["12:00", "13:00", "17:00", "18:00", "18:30", "19:00"])
    ).toBe(2.5);
  });

  test("Invalid break array input", () => {
    expect(calculateBreak(["12:00", "invalid"])).toBe(1); // Assuming 1 is the default break duration
  });
});

/**
 * Calculate the end time based on hours left, break duration, start time, and clock in/out times.
 * @param {number} hoursLeft - The number of hours left for the user to work this week.
 * @param {number} breakDuration - The total duration of breaks in hours.
 * @param {number} start - The start time as a number (e.g., 9.25 for 9:15 AM).
 * @param {string[]|undefined} timeArray - An array containing clock in/out times as strings in the "HH:mm" format.
 * @returns {string} - A message with the end time or an appropriate status message.
 */
describe("endTime", () => {
  test("Still at work: endTime, weeklyhr-break-start-time", () => {
    expect(endTime(8, 1, 8, ["08:00"])).toBe("17:00");
  });

  test("Finished Work", () => {
    expect(endTime(8, 1, 9, ["08:00", "17:00"])).toBe("Have a Nice Evening 🦉");
  });

  test("When on the end of the week", () => {
    expect(endTime(-1, 1, 9, undefined)).toBe("Have a Nice Weekend 🚀");
  });

  test("When clocked in twice in one day", () => {
    expect(endTime(8, 1, 9, ["08:00", "17:00", "18:00"])).toBe(
      "🙇‍♂️"
    );
  });

  test("Negative hours", () => {
    expect(endTime(-2, 1, 9, undefined)).toBe("Have a Nice Weekend 🚀");
  });
});

/**
 * Format a number to ensure it has two digits (e.g., 8 -> "08").
 * @param {number} number - A number to format.
 * @returns {string|number} - The formatted number as a string or the original number if it has two or more digits.
 */
describe("singleDigits", () => {
  test("returns 24hrs for single digits", () => {
    expect(singleDigits(8)).toBe("08");
  });

  test("returns 24hrs for double digits", () => {
    expect(singleDigits(12)).toBe(12);
  });

  test("returns 00 for 0", () => {
    expect(singleDigits(0)).toBe("00");
  });

  test("returns 59 for 59", () => {
    expect(singleDigits(59)).toBe(59);
  });
});

describe("Performance tests", () => {
  test("startTime performance", () => {
    const startTimeArray = ["09:15"];
    const defaultStart = 8;

    const startTimeBefore = performance.now();
    startTime(startTimeArray, defaultStart);
    const startTimeAfter = performance.now();

    const elapsedTime = startTimeAfter - startTimeBefore;
    expect(elapsedTime).toBeLessThan(50); // Make sure the function executes in less than 50 milliseconds
  });
});


Mr_Wolf/background.js
/**
 * Listen for a message to set default values in storage
 * @function
 * @param {Object} request - The message payload
 * @param {Object} sender - Details about the sender of the message
 * @param {Function} sendResponse - A function to send a response back to the sender
 * @returns {boolean} true if the message was handled
 */
chrome.runtime.onMessage.addListener(setDefaultValues);

/**
 * Listen for a message to send default values to the active tab
 * @function
 * @param {Object} request - The message payload
 * @param {Object} sender - Details about the sender of the message
 * @param {Function} sendResponse - A function to send a response back to the sender
 * @returns {boolean} true if the message was handled
 */
chrome.runtime.onMessage.addListener(sendDefaultValuesIfOnSmaregi);

/**
 * Handles the "default_values" message to set default values in storage
 * @function
 * @param {Object} request - The message payload
 * @param {Object} sender - Details about the sender of the message
 * @param {Function} sendResponse - A function to send a response back to the sender
 * @returns {boolean} true if the message was handled
 */
function setDefaultValues(request, sender, sendResponse) {
  if (request.action === "default_values") {
    const { hours, start, clicked } = request;
    chrome.storage.sync.set({ hours, start, clicked }, () => {
    });
    return true;
  }
}

/**
 * Handles the "on_smaregi" message to send default values to the active tab
 * @function
 * @param {Object} request - The message payload
 * @param {Object} sender - Details about the sender of the message
 * @param {Function} sendResponse - A function to send a response back to the sender
 * @returns {boolean} true if the message was handled
 */
function sendDefaultValuesIfOnSmaregi(request, sender, sendResponse) {
  if (request.action === "on_smaregi") {
    sendDefaultValues();
    return true;
  }
}

/**
 * Sends default values to the active tab if it matches the smaregi URL pattern
 * @function
 */
function sendDefaultValues() {
  chrome.storage.sync.get(["hours", "start", "clicked"], ({ hours, start, clicked}) => {
    chrome.tabs.query({ active: true, currentWindow: true }, (tabs) => {
      const tab = tabs.find((tab) =>
        tab.url?.match("https://timecard1.smaregi.jp/staffs/dashboard/*")
      );
      if (tab) {
        const { id } = tab;
        chrome.tabs.sendMessage(id, {
          hours,
          start,
          clicked,
          action: "back_default_values",
        });
      }
    });
  });
}


Mr_Wolf/scripts/popup.js
/**
 * Sets default values for hours and start time input fields and sends them to a background service worker.
 * @function
 * @returns {void}
 */
window.onload = () => {
  // Get references to the input fields for hours and start time
  const hoursInput = document.querySelector("#hours");
  const startInput = document.querySelector("#start-time");

  // Set default values for input fields if not already set by user
  hoursInput.value = localStorage.getItem("hours") || 40;
  startInput.value = localStorage.getItem("start") || 8;

  // Get reference to the submit button and add an event listener
  const submitButton = document.querySelector("#submit");
  submitButton.addEventListener("click", (event) => {
    event.preventDefault();
    // Save default values to local storage
    saveDefaultValues();
    // Send default values to the background service worker
    sendValuesToBackground();
  });

  /**
   * Saves default values for hours and start time to local storage.
   * @function
   * @returns {void}
   */
  function saveDefaultValues() {
    localStorage.setItem("hours", hoursInput.value);
    localStorage.setItem("start", startInput.value);
  }

  /**
   * Sends default values for hours and start time to the background service worker.
   * @function
   * @returns {void}
   */
  function sendValuesToBackground() {
    const hours = localStorage.getItem("hours");
    const start = localStorage.getItem("start");
    const clicked = Date.now();
    // Send message to the background service worker with default values and action type
    chrome.runtime.sendMessage({
      hours,
      start,
      clicked,
      action: "default_values",
    });
  }
};

document.getElementById("submit").addEventListener("click", function () {
  const spinner = document.getElementById("spinner");
  const submit = document.getElementById("submit");
  submit.style.display = "none";
  spinner.style.display = "inline-block";

  setTimeout(() => {
    spinner.style.display = "none";
    submit.style.display = "inline-block";
    window.close();
  }, 1000); // Set the duration for which the spinner will be displayed (e.g., 1 seconds)
});


Mr_Wolf/scripts/index.js
// uncomment the following line to use Jest testing
// import { chrome } from "jest-chrome";
const testing = false;

function debug(name, value) {
  testing ? console.log(`${name}: `, value) : null;
}

window.onload = () => {
  run();
};

/**
 * Runs the main function for calculating and displaying time-related information
 */
function run() {
  const today = new Date().getDay();

  const weeklyHours = parseInt(window.localStorage.getItem("hours")) || 40;
  const defaultStart = parseInt(window.localStorage.getItem("start")) || 8;

  const startAndEndTime = document
    .querySelectorAll(".cal_table_detail")
    [today].querySelector(".total.text-middle")
    ?.innerText.match(/\d\d.\d\d/gi);
  debug("startAndEndTime", startAndEndTime);

  const timeWorked = parseFloat(
    document.querySelector("span.data").innerText.match(/.\d.\d\d/)[0]
  );
  debug("timeWorked", timeWorked);

  const daysWorking = Array.from(
    document.querySelectorAll("div.cal_day div.inner")
  )
    .splice(1, 5)
    .filter((e) => e.children.length > 2).length;

  const daysLeftThisWeek =
    startAndEndTime?.length % 2 === 0 || !startAndEndTime
      ? 5 - daysWorking
      : 6 - daysWorking;
  debug("daysLeftThisWeek", daysLeftThisWeek);

  const timeLeftThisWeek = timeLeft(weeklyHours, timeWorked);
  debug("timeLeftThisWeek", timeLeftThisWeek);

  var weeklyHoursDividedByDays = timeLeftThisWeek / daysLeftThisWeek;
  debug("weeklyHoursDividedByDays Before", weeklyHoursDividedByDays);

  if (weeklyHoursDividedByDays == Infinity) {
    weeklyHoursDividedByDays = timeLeftThisWeek;
  }

  if (
    weeklyHoursDividedByDays == -Infinity ||
    isNaN(weeklyHoursDividedByDays)
  ) {
    weeklyHoursDividedByDays = 0;
  }

  if (!weeklyHoursDividedByDays) {
    weeklyHoursDividedByDays = 0;
  }

  debug("weeklyHoursDividedByDays After", weeklyHoursDividedByDays);

  // We have the day of the week at the top and use it to get the div with todays time.
  const startTimeInFloat = startTime(startAndEndTime, defaultStart);
  debug("startTimeInFloat", startTimeInFloat);

  // ---------------------------------------------------------------------------

  var breakDuration = 1;
  const breakElements = document.querySelectorAll(".cal_table_detail")[today];

  if (breakElements) {
    var breakArray = Array.from(
      document
        .querySelectorAll(".cal_table_detail")
        [today].querySelectorAll(".rest.text-middle")
    )
      .flatMap((e) => e.innerText.match(/\d\d.\d\d/g) || [])
      .filter((e) => e !== null);
  } else {
    var breakArray = [];
  }

  breakDuration = calculateBreak(breakArray);
  debug("breakDuration", breakDuration);

  let breakDurationString = "";
  if (breakArray.length !== 0) {
    breakDurationString = `: <strong>Break: </strong> ${numberToTime(
      breakDuration
    )}`;
  }

  var endTimeInString = endTime(
    weeklyHoursDividedByDays,
    breakDuration,
    startTimeInFloat,
    startAndEndTime,
    daysLeftThisWeek
  );
  debug("endTimeInString", endTimeInString);

  /**
   * Helper function to insert the display of time-related information into the HTML
   * @param {number} timeLeftThisWeek - the amount of time left to work this week
   * @param {number} weeklyHoursDividedByDays - the number of hours divided by days left this week
   * @param {string} endTimeInString - the calculated end time for the day
   * @param {string} breakDurationString - the formatted string for the break duration (if applicable)
   */
  insertDisplay(
    timeLeftThisWeek,
    weeklyHoursDividedByDays,
    endTimeInString,
    breakDurationString
  );
}

// Helper Functions ----------------------------------------------------------

function insertDisplay(
  timeLeftThisWeek,
  weeklyHoursDividedByDays,
  endTimeInString,
  breakDurationString
) {
  if (document.getElementById("mr-wolf-summary")) {
    document.getElementById(
      "mr-wolf-summary"
    ).innerHTML = `<strong>Hours left: </strong> ${numberToTime(
      timeLeftThisWeek
    )}/${numberToTime(
      weeklyHoursDividedByDays
    )} : <strong>Ending Time: </strong>${endTimeInString} ${breakDurationString}`;
  } else {
    const newSpan = document.createElement("span");
    newSpan.innerHTML = `<strong>Hours left: </strong> ${numberToTime(
      timeLeftThisWeek
    )}/${numberToTime(
      weeklyHoursDividedByDays
    )} : <strong>Ending Time:</strong> ${endTimeInString} ${breakDurationString}`;
    newSpan.style = "margin: 0 0 0 1rem; cursor: default;";
    newSpan.setAttribute("id", "mr-wolf-summary");

    const newStyle = document.createElement("style");
    newStyle.innerHTML = `@media screen and (max-width: 768px) {
      #mr-wolf-summary {
          display: block;
          margin: 1rem 0 0 0 !important;
      }
    }`;

    document
      .getElementById("btn_cal_check_show_summary_mode")
      .insertAdjacentElement("afterend", newSpan);
    document
      .getElementById("btn_cal_check_show_summary_mode")
      .insertAdjacentElement("afterend", newStyle);
  }
}

/**
 * Calculates the time left to work this week
 * @param {number} weeklyHours - the total number of hours to work in a week
 * @param {number} timeWorked - the amount of time already worked this week
 * @returns {number} - the time left to work this week
 */
function timeLeft(weeklyHours, timeWorked) {
  const calculatedTime = weeklyHours - timeWorked.toFixed(2);
  if (calculatedTime < 0 || calculatedTime === Infinity) {
    return 0.0;
  }

  return weeklyHours - timeWorked;
}

/**
 * Calculates the start time for the current day
 * @param {array} startAndEndTime - the array of start and end times for the current day
 * @param {number} defaultStart - the default start time (in hours) for the day
 * @returns {number} - the start time for the day
 */
function startTime(startAndEndTime, defaultStart) {
  // If day hasn't started or day is over we return the default start time.
  if (startAndEndTime == undefined) {
    return defaultStart;
  }

  // We then get the start time of today.
  let start = startAndEndTime[0];

  return isValidTime(start) ? timeToNumber(start) : defaultStart;
}

/**
 * Calculates the total break duration for the day
 * @param {array} breakArray - the array of break start and end times for the day
 * @returns {number} - the total break duration for the day
 */
function calculateBreak(breakArray) {
  if (!breakArray || breakArray.length === 0) return 1;

  // Discard last value if array has odd number of values
  if (breakArray.length % 2 === 1) {
    breakArray.pop();
  }

  const breaks = [];

  for (let i = 0; i < breakArray.length; i += 2) {
    const startBreak = breakArray[i];
    const endBreak = breakArray[i + 1];

    if (isValidTime(startBreak) && isValidTime(endBreak)) {
      const start = timeToNumber(startBreak);
      const end = timeToNumber(endBreak);
      breaks.push(end - start);
    }
  }

  return breaks.length > 0 ? breaks.reduce((a, b) => a + b) : 1;
}

/**
 * Checks if a given time string is valid
 * @param {string} timeStr - the time string to check
 * @returns {boolean} - true if the time string is valid, false otherwise
 */
function isValidTime(timeStr) {
  const timeRegex = /^([0-1][0-9]|2[0-3]):[0-5][0-9]$/;
  return timeRegex.test(timeStr);
}

/**
 * Converts a time string to a decimal number
 * @param {string} time - the time string to convert
 * @returns {number} - the decimal representation of the time string
 */
function timeToNumber(time) {
  const [hours, minutes] = time.split(":");
  return parseInt(hours, 10) + parseInt(minutes, 10) / 60;
}

/**
 * Converts a decimal number to a time string
 * @param {number} time - the decimal number to convert
 * @returns {string} - the time string representation of the decimal number
 */
function numberToTime(time) {
  const hours = Math.floor(time);
  const minutes = Math.round((time - hours) * 60);

  return `${hours}:${singleDigits(minutes)}`;
}

/**
 * Calculates the end time for the day
 * @param {number} weeklyHoursDividedByDays - the number of hours divided by days left this week
 * @param {number} breakDuration - the total break duration for the day
 * @param {number} startTimeInFloat - the start time for the day (in decimal form)
 * @param {array} startAndEndTime - the array of start and end times for the day
 * @param {number} daysLeftThisWeek - the number of days left to work this week
 * @returns {string} - the formatted end time for the day
 */
function endTime(
  weeklyHoursDividedByDays,
  breakDuration,
  startTimeInFloat,
  startAndEndTime,
  today
) {
  if (weeklyHoursDividedByDays <= 0 || today == 0 || today == 6) {
    breakDuration = "";
    return "Have a Nice Weekend 🚀";
  }

  if (startAndEndTime?.length % 2 == 0) {
    return "Have a Nice Evening 🦉";
  }

  if (startAndEndTime?.length > 2) {
    return "🙇‍♂️";
  }

  var endTimeInFloat =
    weeklyHoursDividedByDays + breakDuration + startTimeInFloat;

  return numberToTime(endTimeInFloat);
}

/**
 * Adds a leading 0 to a single-digit number
 * @param {number} params - the number to add a leading 0 to
 * @returns {string} - the number with a leading 0 (if applicable)
 */
function singleDigits(params) {
  if (params < 10) {
    return "0" + params;
  }
  return params;
}

// -----------------------------------------------------------------------------

// Checks if the user is on smaregi timecard and sends a message to background.js to run sendDefaultValues()
const check_url = setInterval(function () {
  if (window.location.href.includes("timecard1.smaregi.jp/staffs/dashboard")) {
    if (document.querySelectorAll(".week.cf.show_shift").length !== 1) {
      if (document.getElementById("mr-wolf-summary")) {
        document.getElementById("mr-wolf-summary").innerHTML =
          "Change to shift view to view summary";
      } else {
        const newDiv = document.createElement("span");
        newDiv.innerHTML = "Change to shift view to view summary";
        newDiv.style.margin = "1rem";
        newDiv.setAttribute("id", "mr-wolf-summary");

        document
          .getElementById("btn_cal_check_show_summary_mode")
          .insertAdjacentElement("afterend", newDiv);
      }
      return;
    }

    chrome.runtime.sendMessage({
      action: "on_smaregi",
    });
  }
}, 1000);

check_url;

// Gets the credentials from service worker(background.js) and saves them in session storage
chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
  if (request.action === "back_default_values") {
    let back_hours = request.hours;
    let back_start = request.start;
    let back_clicked = request.clicked;

    if (back_clicked > Date.now() - 1000) {
      window.localStorage.setItem("hours", back_hours);
      window.localStorage.setItem("start", back_start);
      run();
    }

    if (
      back_hours != window.localStorage.getItem("hours", back_hours) ||
      back_start != window.localStorage.getItem("start", back_start)
    ) {
      window.localStorage.setItem("hours", back_hours);
      window.localStorage.setItem("start", back_start);
      run();
    }
  }
});

// define the callback function to be executed when changes are detected
const observerCallback = function (mutationsList) {
  for (const mutation of mutationsList) {
    if (
      mutation.type === "childList" ||
      mutation.target.classList.contains("week")
    ) {
      run();
      break;
    }
  }
};

// create a new mutation observer and attach it to the #cal_table element
const targetNode = document.querySelector("#cal_table");
const observer = new MutationObserver(observerCallback);
const observerOptions = { childList: true, subtree: true, attributes: true };
observer.observe(targetNode, observerOptions);

// uncomment the following line to use Jest testing

// module.exports = {
//   timeLeft: timeLeft,
//   startTime: startTime,
//   calculateBreak: calculateBreak,
//   endTime: endTime,
//   singleDigits: singleDigits,
// };
